ggml/src/ggml-qnn/dl-loader.hpp

```diff
@@ -0,0 +1,76 @@
+#pragma once
+
+#ifdef __linux__
+#    include <dlfcn.h>
+#    include <fcntl.h>
+#elif defined(_WIN32)
+#    define WIN32_LEAN_AND_MEAN
+#    ifndef NOMINMAX
+#        define NOMINMAX
+#    endif
+#    include <windows.h>
+#endif
+
+#include <string>
+
+namespace qnn {
+
+#ifdef __linux__
+typedef void * dl_handler_t;
+
+inline qnn::dl_handler_t dl_load(const std::string & lib_path) {
+    return dlopen(lib_path.c_str(), RTLD_NOW | RTLD_LOCAL);
+}
+
+inline void * dl_sym(qnn::dl_handler_t handle, const std::string & symbol) {
+    return dlsym(handle, symbol.c_str());
+}
+
+inline bool dl_unload(qnn::dl_handler_t handle) {
+    return dlclose(handle) == 0;
+}
+
+inline const char * dl_error() {
+    return dlerror();
+}
+#elif defined(_WIN32)
+using dl_handler_t = HMODULE;
+
+inline qnn::dl_handler_t dl_load(const std::string & lib_path) {
+    // suppress error dialogs for missing DLLs
+    auto old_mode = SetErrorMode(SEM_FAILCRITICALERRORS);
+    SetErrorMode(old_mode | SEM_FAILCRITICALERRORS);
+
+    auto handle = LoadLibraryA(lib_path.c_str());  // TODO: use wstring version for unicode paths
```

Comment:
Hi @slaren, noticed we have similar dynamic library loading functionality in[ggml-bacnend-reg.cpp](https://github.com/ggml-org/llama.cpp/blob/4375415b4abf94fb36a5fd15f233ac0ee23c0bd1/ggml/src/ggml-backend-reg.cpp#L106) (the `dl_load_library` function) that could be useful in other parts of the codebase. 
I suggest moving this to a common utility module so we can reuse it across the project. This would help reduce code duplication and provide a consistent approach to loading libraries.
I'd be happy to prepare another PR about that.

Reply:
Sorry, I missed this. I think that this code is small enough that it is not really a problem if it is duplicated in a backend, and making it part of the public API available to backends may make it harder to change it in the future. So at the moment my preference would be to avoid this.